gms/gossiper.cc

```diff
@@ -775,6 +730,84 @@ future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep) {
     });
 }
 
+
+future<> gossiper::failure_detector_thread_for_node(gms::inet_address node) {
+    auto last = gossiper::clk::now();
+    auto max_duration = std::chrono::milliseconds(_cfg.failure_detector_timeout_in_ms());
+    auto diff = std::chrono::milliseconds(0);
+    auto nr = _live_endpoints.size();
+    while (is_enabled()) {
+        bool failed = false;
+        try {
+            logger.debug("failure_detector_thread: Send echo to node {}", node);
+            co_await _messaging.send_gossip_echo(netw::msg_addr(node), max_duration);
+            logger.debug("failure_detector_thread: Send echo to node {}, status = ok", node);
+        } catch (...) {
+            failed = true;
+            logger.info("failure_detector_thread: Send echo to node {}, status = failed: {}", node, std::current_exception());
+        }
+        auto now = gossiper::clk::now();
+        diff = now - last;
+        if (!failed) {
+            last = now;
+        }
+        if (diff > max_duration) {
+            logger.info("failure_detector_thread: Mark node {} as DOWN", node);
+            co_await container().invoke_on(0, [node] (gms::gossiper& g) {
+                return seastar::async([node, &g] {
+                    g.convict(node);
+                });
+            });
+            co_return;
+        } else if (_live_endpoints.size() != nr) {
+            logger.info("failure_detector_thread: Finish loop for node {}, live_endpoints={}", node, _live_endpoints);
+            co_return;
+        } else  {
+            co_await sleep_abortable(std::chrono::milliseconds(2000), _abort_source);
+        }
+    }
+    co_return;
+
+}
+
+future<> gossiper::failure_detector_thread() {
+    auto shard = this_shard_id();
+    logger.info("failure_detector_thread: Started on shard={}", shard);
+    while (is_enabled()) {
+        try {
+            utils::chunked_vector<inet_address> nodes;
+            while ((_live_endpoints.size() <= shard) && is_enabled()) {
+                logger.info("failure_detector_thread: Wait for live nodes to work with on shard={}, live_nodes={}, live_nodes_for_this_shard={}",
+                        shard,  _live_endpoints, nodes);
+                co_await sleep_abortable(std::chrono::milliseconds(1000), _abort_source);
+            }
+            nodes = co_await container().invoke_on(0, [shard] (gms::gossiper& g) {
+                utils::chunked_vector<inet_address> nodes;
+                for (unsigned idx = 0; idx < g._live_endpoints.size(); idx++) {
+                    if (idx % smp::count == shard) {
+                        nodes.push_back(g._live_endpoints[idx]);
+                    }
+                }
+                return nodes;
+            });
+            if (nodes.empty()) {
+                logger.info("failure_detector_thread: Got empty live nodes to work on shard={}, live_nodes={}, live_nodes_for_this_shard={}",
+                        shard, _live_endpoints, nodes);
```

Comment:
Too much noise.

Reply:
OK. I will make it debug.

Reply:
There's plenty of logger.info calls, please check them all.